Add images chapter to classic-ui docs

MrTango opened the pull request.

MrTango committed:
Add images chapter to classic-ui docs

mister-roboto commented:
@MrTango thanks for creating this Pull Request and helping to improve Plone!

TL;DR: Finish pushing changes, pass all other checks, then paste a comment:
```
@jenkins-plone-org please run jobs
```

To ensure that these changes do not break other parts of Plone, the Plone test suite matrix needs to pass, but it takes 30-60 min.  Other CI checks are usually much faster and the Plone Jenkins resources are limited, so when done pushing changes and all other checks pass either [start all Jenkins PR jobs yourself](https://jenkinsploneorg.readthedocs.io/en/latest/run-pull-request-jobs.html#run-a-pull-request-job), or simply add the comment above in this PR to start all the jobs automatically.

Happy hacking!

MrTango committed:
Add note about permissions in ImageScaling

stevepiercy commented on docs/classic-ui/images.md:
```diff
@@ -0,0 +1,114 @@
+        scale=None,
+        height=None,
+        width=None,
+        direction="thumbnail",
```
If `direction` is deprecated, should we include it? Should we replace it with `mode`?

stevepiercy replied:
Also this is duplicated from earlier.

jensens replied:
Docs should use `mode`.

MrTango replied:
right now, only plone.scaling implements mode, plone.namedfile which is the main package we are dealing with here, does not yet ;). So we have to use the old direction parameter until, this is refactured.

stevepiercy replied:
OK. That makes sense by itself. However the section "Scaling `direction` deprecated in favor of `mode`" creates confusion. It appears to override everything written up to that point.

We could add [`version-changed`](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-versionchanged) and [`deprecated`](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-deprecated) directives to make it clear.

MrTango replied:
I think we either refacture it right away or we don't mention the chnage for now, as the api of plone.namedfile has not changed yet. The rest is internal for now.

stevepiercy replied:
Let's put this section in an `attention` admonition. That will both save the content for when y'all get around to refactoring it and notify users that this change is coming. I think we can also tune the language to make it clear that it is not ready yet.

stevepiercy replied:
I pushed a commit https://github.com/plone/documentation/pull/1192/commits/58481e400f28d62b1ad48bdf33572b2710c6f81c that should address this.

stevepiercy commented on docs/classic-ui/images.md:
```diff
@@ -0,0 +1,114 @@
+
+# Image resolving and scaling
+
+Image scaling is done in plone.namedfile.scaling and most interesting the ImageScaling view.
```
Where is "ImageScaling view"? What is it? This might be obvious to experienced Plone developers, but newbies have no idea.

jensens replied:
Precondition: A (Dexterity) content type with one or more image fields (`plone.namedfile.NamedBlobImage`). `context` is an instance of this content type.

View is accessed via traversal: 
`context.restrictedTraverse("@@images/FIELDNAME")` to get the view for s specific field.

The (by default registered) view class is `plone.namedfile.scaling.ImageScaling`.

The view is 
- on one hand a tool for programmers to get reliable and cachable URLs of scales of an image of above mentioned image field, 
- on the other hand it is used as a URL in the browser like `../my.jpg/@@images/SCALEUID` or (not cachable and not recommended) like `../my.jpg/@@images/FIELDNAME/SCALENAME`.

jensens replied:
I just found plone.api is the way to go, so forget about the traversal part. Only needed in plone core development where plone.api can not be used.

MrTango replied:
restrictedTraverse is not working for this case, it's not able to resolve the image field and scale.

stevepiercy replied:
I don't think I got an answer to my original question: Where is "ImageScaling view"? What is it?

MrTango replied:
it's linked now

stevepiercy commented on docs/classic-ui/images.md:
```diff
@@ -0,0 +1,114 @@
+keep | scale
+thumbnail | scale
+
+For now plone.namedfile is still expecting the direction argument with the old values.
```
Are we supposed to use `direction="foo"` or `mode="foo"`? Also effective in which version of Plone or `plone.api`?

jensens replied:
`mode` is the way to go. `direction` should be mentioned somewhere, it's in for backward compatibility reasons.

jensens replied:
`mode` is available in `plone.scale` since version 3.1.0. This is not specific to `plone.api`.

stevepiercy replied:
This was resolved in https://github.com/plone/documentation/pull/1192/commits/58481e400f28d62b1ad48bdf33572b2710c6f81c

stevepiercy commented on docs/classic-ui/images.md:
```diff
@@ -0,0 +1,114 @@
+
+## Permissions
+
+The ImageScaling view is checking explicitly the permissions the current user has, in case you want to access objects which are normally not accessible to the current user, you have to override the validate_access method in ImageScale. In Products.EasyNewsletter you can find an example of that.
```
"In Products.EasyNewsletter you can find an example of that."

Please add hyperlink to reference.

stevepiercy replied:
I searched for [`ImageScaling` in `Products.EasyNewsletter`](https://github.com/collective/Products.EasyNewsletter/search?q=ImageScaling), but came up empty. Where is the example?

MrTango replied:
i will leave it out for now, i have to make soem testing there, to make sure it is save to use. But it is in src/Products/EasyNewsletter/views/newsletter_issue_send.py, right now it's just disabling the check, which is not good. We need more control here, i want to disable it inside the view, but it should not allow anonoumus access outside that view for every image ;)

stevepiercy committed:
English grammar fixes
Surround inline literals with backticks
Add html_meta tags
Add labels for future navigation ease

stevepiercy reviewed (changes requested):
Thanks for getting started on this!

I made several changes in facc095bcb543d2961f30e3f2fe4a6424f73dea0 and pushed them to this PR. Most of them were just English grammar or MyST syntax. However I was unsure about a couple of technical points and terminology, so they should be reviewed for accuracy.

I also asked several questions in the review. Please respond at your convenience.

jensens commented on docs/classic-ui/images.md:
```diff
@@ -0,0 +1,135 @@
+from plone import api
+
+scale_util = api.content.get_view("images", context, request)
+image_scale = scale_util.scale("leadimage", scale="mini")
```
```suggestion
scale_util = api.content.get_view("images", context, request)
# on the following line "leadimage" is the field name. The default Image content types field name is "image".
image_scale = scale_util.scale("leadimage", scale="mini")
```

jensens commented on docs/classic-ui/images.md:
````diff
@@ -0,0 +1,135 @@
+At this point, you can also get an HTML tag, since you have created an instance of `api.content.get_view` in `scale_util`:
+
+```pycon
+>>> print(image_scale.tag)
````
`tag` is a method, so this won't work. see https://github.com/plone/plone.namedfile/blob/master/plone/namedfile/scaling.py#L89
```suggestion
>>> print(image_scale.tag())
```

jensens replied:
Further it is important to document the various options of tag.

jensens commented on docs/classic-ui/images.md:
```diff
@@ -0,0 +1,135 @@
+from plone import api
+
+scale_util = api.content.get_view("images", context, request)
+tag = scale_util.scale("leadimage", width="600", height="200")
```
One step is missing.
```suggestion
image_scale = scale_util.scale("leadimage", width="600", height="200")
image_scale.tag()

# or by shortcut
tag = scale_util.tag("leadimage", width="600", height="200")
```

stevepiercy replied:
@MrTango what do you think about this one?

MrTango replied:
outdated

jensens commented on docs/classic-ui/images.md:
```diff
@@ -0,0 +1,135 @@
+-   `url`
+-   `width`
+
+At this point, you can also get an HTML tag, since you have created an instance of `api.content.get_view` in `scale_util`:
```
The sentence twists my brain. Just:
```suggestion
You can directly create an HTML tag from `image_scale`:
```

jensens commented on docs/classic-ui/images.md:
````diff
@@ -0,0 +1,135 @@
+<img src="http://localhost:8080/Plone/news/newsitem1/@@images/9f676d46-0cb3-4512-a831-a5db4079bdfa.jpeg" alt="News Item 1!" title="News Item 1" height="21" width="32" srcset="http://localhost:8080/Plone/news/newsitem1/@@images/4a68513c-cffd-4de0-8a35-80627945b80f.jpeg 2x, http://localhost:8080/Plone/news/newsitem1/@@images/c32929c6-cb89-4ce7-846f-38adf29c09a4.jpeg 3x" />
+```
+
+You can also get an HTML tag with a specific size:
````
```suggestion
Instead of using the configured named scales you can also get an HTML tag with any specific size in pixels:
```

jensens commented on docs/classic-ui/images.md:
```diff
@@ -0,0 +1,135 @@
+
+## Permissions
+
+The `ImageScaling` view explicitly checks the permissions of the current user.
```
```suggestion
The `ImageScaling` view `@@images` is protected by the `View` permission.
```

jensens commented on docs/classic-ui/images.md:
```diff
@@ -0,0 +1,135 @@
+## Permissions
+
+The `ImageScaling` view explicitly checks the permissions of the current user.
+If you want to access objects which are normally not accessible to the current user, then you have to override the `validate_access` method in `ImageScale`.
```
```suggestion
To access image scales which are normally not accessible to the current user override the `validate_access` method in `plone.namedfile.scaling.ImageScale`.
```

jensens reviewed (changes requested):
Needs some fine tuning.

stevepiercy committed:
Fix Sphinx warning from duplicate label

MrTango committed:
Update docs/classic-ui/images.md

Co-authored-by: Jens W. Klein <jk@kleinundpartner.at>

MrTango committed:
Update docs/classic-ui/images.md

Co-authored-by: Jens W. Klein <jk@kleinundpartner.at>

MrTango committed:
Update docs/classic-ui/images.md

Co-authored-by: Jens W. Klein <jk@kleinundpartner.at>

MrTango committed:
Update docs/classic-ui/images.md

Co-authored-by: Jens W. Klein <jk@kleinundpartner.at>

MrTango committed:
Update docs/classic-ui/images.md

Co-authored-by: Jens W. Klein <jk@kleinundpartner.at>

stevepiercy committed:
Clean up grammar and punctuation.
Move "Scaling `direction` to be deprecated in favor of `mode`" into an "attention" admonition,
and change wording to be more clear about which form to use and when.

MrTango committed:
Improve image handling docs

MrTango committed:
Merge branch 'mrtango-classic-ui-images' of https://github.com/plone/documentation into mrtango-classic-ui-images

stevepiercy commented on docs/classic-ui/images.md:
```diff
@@ -0,0 +1,252 @@
+  "property=og:title": "Image resolving and scaling"
+  "keywords": "Plone, Classic UI, classic-ui, image, resize, scale"
+---
+****
```
Not sure where these came from.

```suggestion

```

stevepiercy reviewed (changes requested):
There are weird merge commits in this that should not be there, and the entire page was overhauled.

MrTango committed:
resolve merge conflicts

MrTango committed:
Update docs/classic-ui/images.md

Co-authored-by: Steve Piercy <web@stevepiercy.com>

jensens commented on docs/classic-ui/images.md:
```diff
@@ -0,0 +1,226 @@
+
+### By cacheable scale UID name
+
+When an image scale was created, it will be cached under the name `UID.jpeg` in the object annotations and can be resolved by it as follows:
```
Worth to mention its not limited to jpeg?

stevepiercy replied:
What are the possible image formats?

When an image is resized, is the resized image in `jpeg` format?

jensens replied:
gif stays gif, png stays png, same for jpeg, just tiff gets converted to IIRC jpeg.

mauritsvanrees replied:
- png stays png
- svg stays svg
- gif becomes png
- the rest becomes jpeg

See https://github.com/plone/plone.scale/blob/a5cb08617fcd5abe3ec95f895926428e7551a8f9/plone/scale/scale.py#L63-L68

jensens commented on docs/classic-ui/images.md:
```diff
@@ -0,0 +1,226 @@
+
+http://localhost:8080/Plone/news-item1/@@images/3d182f34-8773-4f20-a79d-8774c3151b7e.jpeg
+
+This very useful for caching URL's in Varnish or the Browser. In case the scale definitions have changed, they will be saved again under a different UID, so that the URL will change and enforce the Browser or a cache proxy like Varnish to fetch it again.
```
```suggestion
This very useful for caching URL's in Varnish or the Browser. In case the uploaded image or scale definitions have changed, they will be saved again under a different UID, so that the URL will change and enforce the Browser or a cache proxy like Varnish to fetch it again.
```

jensens replied:
most important: new image was uploaded.

stevepiercy replied:
Incorporated in b6c77816267f8940973a85010a4f37dc9730f58e

jensens commented on docs/classic-ui/images.md:
```diff
@@ -0,0 +1,226 @@
+from plone import api
+
+scale_util = api.content.get_view("images", context, request)
+tag = scale_util.tag(
+    fieldname=None,
+    scale=None,
+    height=None,
+    width=None,
+    direction="thumbnail"
+)
```
I think here we should mention, additional kwargs are ending up as attributes on the tag.
https://github.com/plone/plone.namedfile/blob/ecf33a2bc7a8c61888909bc383b3e08d80888e43/plone/namedfile/scaling.py#L123

stevepiercy replied:
Incorporated suggestion in b6c77816267f8940973a85010a4f37dc9730f58e

jensens reviewed (approved):
Great work, just minor suggestions.

stevepiercy committed:
Fix MyST syntax, English grammar
Add MyST labels to section headings
Add suggestions from @jensens

stevepiercy commented:
Respond to https://github.com/plone/documentation/pull/1192#pullrequestreview-899494381, then one more round of reviews, and I think we are done! @jensens @MrTango

jensens commented on docs/classic-ui/images.md:
```diff
@@ -0,0 +1,246 @@
+
+### By cacheable scale UID name
+
+When an image scale is created, it will be cached under the name `UID.jpeg` in the object annotations.
+It can be resolved as follows:
```
```suggestion
When an image scale is created, it will be cached under the name `UID.EXT` (i.e. `f4c34254b44ba351af7393bfe0296664.jpeg`) in the object annotations.
Scaling keeps the uploaded formats, except for TIFF which ends up as JPEG.
It can be resolved as follows:
```

MrTango committed:
Update docs/classic-ui/images.md

Co-authored-by: Jens W. Klein <jk@kleinundpartner.at>

MrTango commented:
@stevepiercy should be ok now

MrTango commented:
@mauritsvanrees you where recendly more intomthe scaling directions, would you at some point add some describtions to the scaling directions?

stevepiercy reviewed (approved):
LGTM! Thank you! Please merge whenever you like.

stevepiercy commented:
The one failing test is from a broken link:
https://github.com/plone/documentation/runs/5422538177?check_suite_focus=true#step:5:958

We can ignore that for this PR. The Razzle site might not accept HEAD requests, which linkcheck uses to check for valid links.

mauritsvanrees commented:
> @mauritsvanrees you where recendly more intomthe scaling directions, would you at some point add some describtions to the scaling directions?

I can do that in a later PR.
Also, direction is deprecated and we should use mode. But I think lots of core code still uses direction.

stevepiercy commented:
> > @mauritsvanrees you where recendly more intomthe scaling directions, would you at some point add some describtions to the scaling directions?
> 
> I can do that in a later PR. Also, direction is deprecated and we should use mode. But I think lots of core code still uses direction.

To not reinvent the wheel, this commit has most of the content for deprecation of `direction` in favor of `mode`.

https://github.com/plone/documentation/pull/1192/commits/58481e400f28d62b1ad48bdf33572b2710c6f81c

We discussed whether to include it now or later, and this commit attempted to appease both the current and future state, but ultimately it was removed for now.

MrTango commented:
> Also, direction is deprecated and we should use mode. But I think lots of core code still uses direction.

yes, plone.namedfile still uses the old api and that's what people use most in the end. 
That's why i removed the mentioning for now. After we refactore plone.namefiles, we can update the docs.

@stevepiercy from my point of view this is good to merge, but there are some link test failures, but i don't see why.

MrTango commented:
just saw your comment, ok then i'll merge it now.
